sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -139,6 +139,7 @@ class Analyzer(
       ExtractGenerator ::
       ResolveGenerate ::
       ResolveFunctions ::
+      ResolveLiteralFunctions ::
```

Comment:
The order matters. It assumes `ResolveReferences` should be run before this rule. However, `ResolveReferences` might need multiple passes to resolve all the references. Thus, how about moving the logics into `ResolveReferences `? If the attributes are not resolvable, we try to see whether it is a function literal?

Reply:
Agree! I will refactor it.